Dynamic Dispatch for separately stored karaoke effects

ahuei123456 opened the pull request:
Moves all karaoke effect generators into classes and adds in dynamic dispatch to retrieve effects from files that are not contained in the repo.

Two new modifiers are added: `import` and `kfx`.

- `import` is used to specify what Python files are to be imported into the current execution environment. The directory of the input file passed as a command line argument is automatically added to the Python PATH, so any Python file in the same directory as the input file can be imported. `src/ass/kfx.py` is a file that is not automatically imported, and can be done so via this modifier as an example.
- `kfx` is used to specify the name of the effect to be used for the specific song. If none is specified, then it defaults to `default_live_karaoke_effect`, which is stored in `src/ass/to_ass.py`. All effects must be registered into the `effects` dictionary located in `src/ass/to_ass.py`, of which an example can be seen in `src/ass/kfx.py`.

Other considerations:

- `Effect` could possibly be loaded and stored within `Song` instead of externally handled in `populate_song`. This would also create a new method within `Song` to generate events, which would then be called by `populate_song`.
- All the important data structures should be able to be imported at once by an external file, to simplify imports from an effect that is potentially located externally.
- `NoKaraokeEffect` is still incomplete. I'm not sure what exactly it's supposed to generate right now.

xIceArcher commented:
> src/ass/kfx.py is a file that is not automatically imported, and can be done so via this modifier as an example.

I think the eventual state of the repo should be something like:
- `default_effect.py` contains the class `DefaultLiveKaraokeEffect()` and all associated methods. Ideally, this lives in its own directory too separate from `src/ass` (maybe `src/effect`?)
- This file serves both as an the default effect as well as an example on how to use it (i.e. move all DefaultLiveKaraokeEffect() related classes and methods out of to_ass.py and into its own file, just like an external effect would be, same for NoLyricsEffect() and NoKaraokeEffect())
- `src/ass/to_ass.py` only contains the abstract class definitions
- `src/ass/kfx.py` can be removed from this repo

> All effects must be registered into the effects dictionary located in src/ass/to_ass.py, of which an example can be seen in src/ass/kfx.py.

Nitpicky but not a fan of this essentially global variable approach... Maybe make it a function like `register_effect()` or something for better isolation and mark the global effects dict as internal (`_effects`) (although I know technically nothing's ever actually private in Python...)

> Effect could possibly be loaded and stored within Song instead of externally handled in populate_song. This would also create a new method within Song to generate events, which would then be called by populate_song.

Not a fan of this either, since conceptually an effect and a song can (and should) be kept completely separately (in that a `Song` is completely valid and can exist without an `Effect`). 

> All the important data structures should be able to be imported at once by an external file, to simplify imports from an effect that is potentially located externally.

Agreed, I think importing `src/ass` is enough right now?

> NoKaraokeEffect is still incomplete. I'm not sure what exactly it's supposed to generate right now.

No worries, I'll do that one in due time

ahuei123456 commented:
> I think the eventual state of the repo should be something like...

Updated.

> Nitpicky but not a fan of this essentially global variable approach... Maybe make it a function like register_effect() or something for better isolation and mark the global effects dict as internal (_effects) (although I know technically nothing's ever actually private in Python...)

Also updated.

> Not a fan of this either, since conceptually an effect and a song can (and should) be kept completely separately (in that a Song is completely valid and can exist without an Effect).

Hmm, understood. Though, I think there should be a way to specify a global effect for all songs in a particular file (or just default to `default_live_karaoke_effect` always?).

> Agreed, I think importing src/ass is enough right now?

Looks fine in the current effects.

xIceArcher commented:
> Hmm, understood. Though, I think there should be a way to specify a global effect for all songs in a particular file (or just default to default_live_karaoke_effect always?).

Yeah, I think that's good enough. You _could_ potentially add a command-line flag to `populate_songs.py` and allow specifying a different default effect, but to be really honest I don't ever see it being used, so...

xIceArcher commented on src/ass/to_ass.py:
```diff
@@ -1,84 +1,242 @@
+        shouldPrintTitle: bool = True,
+        dividerStyle: pyass.Style = DIVIDER_STYLE,
+        titleStyle: pyass.Style = TITLE_STYLE,
+        end: timedelta = TITLE_EVENT_DURATION,
```
Oops, missed this on the previous round of review

`end` -> `titleEventDuration`

ahuei123456 replied:
Fixed.

xIceArcher commented on src/ass/to_ass.py:
```diff
@@ -1,84 +1,242 @@
-                transitionDuration,
-            )
-            for i, line in enumerate(song.lyrics)
-        ],
-        to_divider_event(song, "English"),
-        *[
-            to_default_event(
-                to_en_k_line(line, i + 1),
-                actorToStyle,
-                switchDuration,
-                transitionDuration,
-            )
-            for i, line in enumerate(song.lyrics)
-        ],
+    switchDuration: timedelta,
+) -> Sequence[pyass.Tag]:
+    if len(kChar.line.actorSwitches) == 0:
+        return []
+    return list(actorToStyle[kChar.line.startActor]) + [
+        pyass.TransformTag(
+            start=time + kChar.fadeOffset,
+            end=time + kChar.fadeOffset,
+            to=actorToStyle[actor],
+        )
+        for time, actor in kChar.line.actorSwitches
     ]
+
+
+def get_char_karaoke_tag(
+    kChar: KChar, resultTag: Sequence[pyass.Tag] = []
+) -> Sequence[pyass.Tag]:
+    if kChar.karaDuration == timedelta():
+        return []
+    # Karaoke timing
+    if len(resultTag) == 0:
+        return [pyass.KaraokeTag(kChar.karaDuration)]
+    else:
+        return [
+            pyass.TransformTag(start=kChar.karaStart, end=kChar.karaEnd, to=resultTag)
+        ]
```
Do you see these methods as belonging here or in `effect/default_effect.py`?

I think things like `get_exit_transition_tag` seem quite specific to the default effect and is unlikely to see any re-use in other user-defined effects, but if you think otherwise, feel free to ignore this comment. 

If the functions are only relevant to `default_effect`, though, I think they can be moved to `default_effect.py` as well?

ahuei123456 replied:
Good point, I can't think of any instance where that would be the case. Updated.

xIceArcher reviewed (commented):
Aside from the two remaining comments, LGTM!
